### DISCUSSION TITLE
Thermal hydraulic model long initialisation times

### DISCUSSION THREAD

**Discussion by jvwilliams23** [idaholab/moose#25555] [Q&A Modules: Thermal Hydraulics] [answered]
Hi,

I am running a thermal hydraulic analysis of a coolant system. The graph has about 808 FlowChannels and 557 VolumeJunctions to connect them. 

The issue I am having is when I begin to run the simulation (using the 'combined' module solver) it takes a long time at the setup stage of "Initializing Equation Systems". To do this, it takes 5000 seconds and uses 814 MB memory. The mesh only has 3232 nodes and 2424 elements. I have noticed this time increases when I use more elements and also more VolumeJunctions (if I use JunctionOneToOne1Phase for all junctions with 1 inlet and 1 outlet, it takes like 300 seconds to initialise, which is still quite long).

Is there any way to avoid this? It seems there is a variable created for every "junction" (also these show up in paraview as individual variables), which must cause the issue. I have attached a log file below.

Best wishes,
Josh
```
--------------------------------------------
Setting Up
  Finished Setting Up Options                                                            [  5.94 s] [  186 MB]


*** Info ***
Projection lowers order, please expect a loss of accuracy
Still Setting Up......
  Initializing
    Initializing Equation Systems........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................... [5027.27 s] [  814 MB]
  Finished Initializing                                                                  [5030.18 s] [ 2804 MB]
Finished Setting Up                                                                      [5079.07 s] [ 2341 MB]

Framework Information:
MOOSE Version:           git commit a643d739e2 on 2023-06-26
LibMesh Version:
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Fri Sep 22 11:07:09 2023
Executable Timestamp:    Tue Sep 12 15:36:10 2023

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          1
  Spatial Dimension:       3
  Nodes:                   3232
  Elems:                   2424
  Num Subdomains:          808

Nonlinear System:
  Num DOFs:                10057
  Num Local DOFs:          10057
  Variables:               { "junction-0:rhoEV" "junction-0:rhoV" "junction-0:rhouV" "junction-0:rhovV"
                             "junction-0:rhowV" ... "junction-99:rhoEV" "junction-99:rhoV" "junction-99:rhouV"
                             "junction-99:rhovV" "junction-99:rhowV" } { "rhoA" "rhoEA" "rhouA" }
  Finite Element Types:    "SCALAR" "MONOMIAL"
  Approximation Orders:    "FIRST" "CONSTANT"

Auxiliary System:
  Num DOFs:                34799
  Num Local DOFs:          34799
  Variables:               "A" "A_linear" { "H" "P_hf" "T" } "T_wall" "e" { "junction-0:T" "junction-0:p"
                             "junction-0:vel" "junction-100:T" "junction-100:p" ... "junction-97:p" "junction-97:vel"
                             "junction-99:T" "junction-99:p" "junction-99:vel" } { "p" "rho" "v" "vel_x"
                             "vel_y" "vel_z" }
  Finite Element Types:    "MONOMIAL" "LAGRANGE" "MONOMIAL" "LAGRANGE" "MONOMIAL" "SCALAR" "MONOMIAL"

  Approximation Orders:    "CONSTANT" "FIRST" "CONSTANT" "FIRST" "CONSTANT" "FIRST" "CONSTANT"

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  TimeIntegrator:          ImplicitEuler
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP (auto)
```

**Comment by GiudGiud** [comment]
Hello

@joshuahansel will have more insights but I think the prime way to accelerate these things is to move from having one object per flow channel and junction to a single vectorized object adding all these components. This will take a little bit of reworking the code, and we typically only do that if we have a need for a faster simulation.

There might also be trivial slow downs of the code. If you profile the code we may just be able to fix them. Please see this page for profiling techniques:
https://mooseframework.inl.gov/application_development/profiling.html

Guillaume

**Reply by jvwilliams23** [reply]
Hi

I tried PerfGraph to profile, it just said the whole time is on EquationSystems::Init. I found this pretty much just calls libMesh. I then profiled using gperftools. Showing the top 12 most time consuming actions below:  

![Screenshot 2023-09-25 at 11 46 48](https://github.com/idaholab/moose/assets/48445365/f4477859-dc04-4b31-b853-25fa838fc289)

Josh

**Reply by GiudGiud** [reply]
@lindsayad knows this code better than I do.
Is there a way to bypass the API and provide the sparsity pattern?

**Reply by lindsayad** [reply]
It would be helpful to see the call graph as well which you can create with `svg > foo.svg` or `png > foo.png`. We spent quite a bit of time trying to improve this for finite volume simulations in https://github.com/libMesh/libmesh/issues/3422. However, the scalar variables here are another wrinkle as they are not conceptually covered by our `GhostingFunctor::map_type` whose keys are elements. Off the top of my head I do not know at what point we handle the sparsity from scalar variables but @roystgnr probably would. It would be nice to see an annotation of lines in `handle_vi_vj` to know what lines are the expensive ones in there. I'm guessing it's just `sort_row` at the end of the method that is taking all the time

**Reply by jvwilliams23** [reply]
Hi, here is the png (had issues with svg in github).

![run1](https://github.com/idaholab/moose/assets/48445365/db975a05-e4fa-4da1-975e-d899ea0b6599)

**Reply by lindsayad** [reply]
Can you run
```
list handle_vi_vj
```

**Reply by roystgnr** [reply]
A `SCALAR` variable, for data at a pipe junction?  Is that really how we do it?  A `SCALAR` variable is one whose value is meaningful on *every* element in your mesh (or at least every element in the subdomain on which it's defined).  Five variables for each of a hundred junctions means, for instance, that an EDGE2 connected to one junction wouldn't have an 11x11 element matrix (3 Lagrange variables on each of 2 nodes, plus 5 for the neighboring junction), it would have a 506x506 element matrix.

If that's really what we want to do, there's probably still room to optimize it (we could handle the 6x6 and 500x6 and 6x500 blocks separately on each element, then only work on the 500x500 block once at the end?), but I'm skeptical that this is really what we want to do.  I'd have thought we'd want a `NodeElem` for each junction's variables, to get the sparsity pattern right.

**Reply by jvwilliams23** [reply]
@lindsayad  I get:
```
(pprof) list handle_vi_vj
Total: 111.61s
No source information for libMesh::SparsityPattern::Build::handle_vi_vj
```

sorry not much help

**Reply by jvwilliams23** [reply]
@roystgnr Yeah I do not understand why a `SCALAR` variable is needed for each junction.

I am working on MOOSE for a nuclear-related project, but my background is biological flows. I would like to do 1D modelling of flow in biomedical geometries in a future project, but the number of junctions will be over 10,000 easily (maybe approaching 100,000). So with the current VolumeJunction this will not be possible.

**Reply by lindsayad** [reply]
That's ok. You've already gone above and beyond with doing any profiling! It's much appreciated. It might be helpful to have your input file so we can do our own profiling, but I think we can also make significant progress by changing the variable type for the junction. @joshuahansel do you want me to create an issue for this, or do you want to?

**Reply by joshuahansel** [reply]
> do you want me to create an issue for this, or do you want to?

I'd like you to have that honor.

**Reply by joshuahansel** [reply]
@roystgnr So there is some kind of variable type for scalar variables that is not `SCALAR`? We haven't been aware of other options. So with our junctions, we expect there to be coupling to the DoFs on the connected pipe ends (perhaps technically even to the DoFs on almost-boundary elements).

**Reply by joshuahansel** [reply]
> or at least every element in the subdomain on which it's defined

A `SCALAR` is defined on a subdomain?

**Reply by lindsayad** [reply]
you can subdomain restrict it which is a pledge to limit its sparsity contribution

**Reply by lindsayad** [reply]
Conceptually it seems like a Lagrange variable subdomain restricted to a `NodeElem` would be the right thing for your system analysis

**Reply by lindsayad** [reply]
I've opened https://github.com/idaholab/moose/issues/25593

**Reply by GiudGiud** [reply]
Are you thinking of a single subdomain for all junctions?
thousands of subdomains has been reported as slower

**Reply by roystgnr** [reply]
> So there is some kind of variable type for scalar variables that is not SCALAR?

Yes, if you're sneaky about it.  @lindsayad is correct that if you take a Lagrange variable (field-valued, but scalar-valued at any specific point within the domain of the field) and you then restrict its domain to a single point, what's left is effectively scalar-valued.  The catch is that now no *other* points in the mesh have to worry about that scalar; only the ones connected to it do.

> A `SCALAR` is defined on a subdomain?

Any variable defaults to being defined on the whole mesh, but can be restricted to being defined only on a set of subdomains within the mesh.

> Are you thinking of a single subdomain for all junctions?

I wasn't, but that would be sufficient, now that I think about it.  A single Lagrange variable, defined on all junction nodes, gives you one scalar value per `NodeElem`, and those scalar values don't erroneously get coupled together because they're not on different nodes of the same element when they're only defined on one-node elements.

> thousands of subdomains has been reported as slower

This isn't a problem if we can get away with one junction subdomain, but even so, we need to just pin this down and fix the bug.  Do we have an open issue for it yet?

There is no inherent reason why changing some integers should make simulations slower.  I'm guessing the issue is as simple as some code doing a nested loop (once over subdomain ids, then an inner loop over `active_subdomain_*` or `active_local_subdomain_*` element iterators) rather than a single loop over all elements that then tests `subdomain_id()` and applies the subdomain-specific code from there.  The former is `O(N_subdomains*N_elem)` (because predicate iterators iterate over the entire element container, they just skip entries that don't match the predicate); the latter is `O(N_elem)`.

I'm seeing the nested loop idiom in `MeshDiagnosticsGenerator.C` and `MeshRepairGenerator.C`, but those aren't going to be the code paths anyone's complained about.  It's also in `AutomaticMortarGeneration.C`, `MortarData.C`, `BlockToMeshConverterGenerator.C`, `RefineBlockGenerator.C`, `PetscDMMoose.C`, but in all those cases the outer loop is over an object-specific subset of subdomains, not *every* subdomain, so they may be fine.  The **triple** nested loop in `NonlinearSystemBase.C` is terrifying me, but there is an inner `if(...hasActiveNodeElemConstraints...)` there screening off the element iteration, so that code is only a total disaster if for some reason we're adding such constraints to every subdomain.

There could be more subtle cases where the outer and inner loops are separated by a function call so I'm not finding them with a quick grep.  Do we have any input files that replicate the problem?

**Reply by jvwilliams23** [reply]
I can provide an input file in few hours.

**Reply by jvwilliams23** [reply]
Here are two input files. One has 100 edges, 99 junctions and takes approx 70 seconds to initialise on my laptop. The other has 150 edges and 149 junctions, takes approx 300 seconds to initialise on my laptop.

[straight-pipe-100.i.txt](https://github.com/idaholab/moose/files/12741901/straight-pipe-100.i.txt)
[straight-pipe-150.i.txt](https://github.com/idaholab/moose/files/12741902/straight-pipe-150.i.txt)

**Reply by roystgnr** [reply]
> Do we have any input files that replicate the problem?

This was just poor wording on my part - it was silly for me to digress about wanting to replicate a different problem in this ticket, not after we determined that *that* problem wouldn't get in the way of fixing this one.

On the other hand, posting input files for *this* problem was also a good idea; that'll make it much easier to test potential fixes.  And such atrocious scaling... `N` elements, each trying to futz with its own O(N^2) dense matrix of `SCALAR`s ...  With O(N^3) total we'd expect a 1.5x larger problem to be 3.375x longer instead of 1.5x longer to initialize here, which is bad enough (it would imply that 100 junctions is already initializing 10,000x slower than it should be), but your numbers look even worse than *that*!  I hope there's not something going on here I'm still missing...

I'm tempted to tackle this on two fronts... getting rid of the `SCALAR` use here as per #25593 is definitely the right thing to do in THM, but maybe it wouldn't hurt to improve the `SCALAR` handling in libMesh too.  My ideas for that would at least get the THM scaling down to O(N^2), which is still much worse than the proper O(N) but which might be a sufficient quick fix for these small cases.  Any `SCALAR` fix would eventually be mooted by the fixes at the THM level, but there are likely other users with large numbers of `SCALAR` variables who really do need them, so optimization there wouldn't go to waste.

**Reply by joshuahansel** [reply, marked as answer]
Ok I've been working on this idea of using field variables on the junctions. I tried it out on your problem with 3 different methods:

- Current: what we're doing currently, i.e., using scalar variables for junctions.
- New A: using field variables on NodeElems, keeping the same variable names unique to each junction, e.g., `junction1:pressure`.
- New B: using field variables on NodeElems, using common variable names for all junctions, e.g., `pressure`, which is fine because these are block-restricted.

Here are rough timing results in seconds, with the 100-edge case first and 150-edge case in parentheses, and the "?" where I didn't bother to try:

- Current: 110(380) setup, 116(390) total
- New A: 29(?) setup, 240(?) total
- New B: ~0(~0) setup, 5.4(9.8) total

So basically, if you just change to field variables but don't take advantage of the fact that you don't need unique variable names anymore, then you do significantly decrease the setup time, but for some reason, you increase the solve time (including residual and Jacobian computation) significantly, giving a worse total time. However, if you do start using common variable names, then you see that setup time go to basically zero (I didn't really measure it), while your solve time stays very good.

So this is great news. I don't yet have an answer to why "New A" is so terrible. I will need to actually use a profiler to do this (and I do plan to do this - maybe it will shed light on other performance pitfalls). But whatever the reason, THM has a good solution here, and I'll work on getting this merged soon.

Thank you to @roystgnr and @lindsayad for the critical suggestions, and thank you @jvwilliams23 for raising the issue and providing good input and examples! I'm hopeful that THM's performance for large problems will be greatly improved! 📈 🎉

**Reply by jvwilliams23** [reply]
Sounds great @joshuahansel. Could you point me to your branch with this code? I would like to try it out

**Reply by joshuahansel** [reply]
Ok I started https://github.com/idaholab/moose/pull/28514 as a WIP. I'm just finishing up fixing tests.

Note that it might require some prerequisite work: the new variable names in the system can cause different ordering of the variables in the nonlinear system, which can lead to convergence issues (https://github.com/idaholab/moose/issues/28503). All in all, I'm hopeful that everything should be merged by the end of next week 🤞

**Comment by joshuahansel** [comment]
@licharlot You have some experience with large numbers of components. We've already applied that dependency resolution fix, right?

**Reply by licharlot** [reply]
Yes, the dependency resolution fix is in (otherwise, that problem would take several days or weeks to initialize). I did use large numbers of channels, but never with a very large number of volume junctions. The increased number of variables is probably slowing things down.

**Reply by jvwilliams23** [reply]
Hi @licharlot. Yes definitely the number of variables. What did you use to connect the channels if not VolumeJunctions? I need to impose some source term at the connections for form loss. The ElbowPipe1Phase is not really suitable on geometries with complex bends.

**Reply by licharlot** [reply]
I was modeling a full reactor core, so I had a large number of parallel channels connected together at each end using a volume junction. So large number of channels, but only 2 huge volume junctions. Could you use the `FormLossFromFunction1Phase` component to apply the form losses? Another possible way would be to use a fancy friction factor.

**Reply by jvwilliams23** [reply]
Thanks! For one-to-one junctions I will just use a higher friction factor on the adjacent pipes as a form loss for now (will move to FormLossFromFunction1Phase later).
So the number of junctions is now approximately 68, and initialisation takes 100 seconds.  

When I increase the number of elements per flowchannel from 3 to 20, it takes around 600 seconds.

**Reply by joshuahansel** [reply]
And to be clear, JunctionParallelChannels1Phase has a form loss option, so that can be used for any case where the connected channels are parallel.

**Reply by NorthMagic** [reply]
I have a similar problem with THM initialization out of time especailly with multi-cpu cores/threads parallel computing. I tried to simulate a full core with hundreds of parallel 1D flow channels and without junctions, it takes more than 300 s to finish the initialization. Moreover, I found that the initialization time increased greatly with the number of used cpu cores and threads, which was unacceptable for doing simulations.
From my understanding to the mentioned answers above, a long-time initialization was cuased by too many variables which did not exist in my simulation.  More impotantly, the initialization time increased with used cpu cores. Could you please give me some suggestions?

![image](https://github.com/idaholab/moose/assets/63328077/59cded32-8c97-413f-897c-432b7db01dc7)

**Reply by joshuahansel** [reply]
Ok so you're saying your simulation does not have junctions, so you shouldn't have the same issue. Can you also profile your simulation? https://mooseframework.inl.gov/application_development/profiling.html

**Reply by NorthMagic** [reply]
Sorry, I have never used this profiling tool. I need more time to install and profile my simulation. I put some new testing outputs in Discussion #27941 and the input file for running THM was posted.
